search-app/src/cmr/search/services/content_service.clj

```diff
@@ -0,0 +1,153 @@
+  content)."
+  {:default 400
+   :type Long})
+
+(defconfig job-wait
+  "Number of milliseconds the `static-content-generator-job` waits for the
+  collection cache to be populated in the event when the delay is not long
+  enough."
+  {:default (* 60 1000)
+   :type Long})
+
+(defconfig retry-count
+  "Number of times `static-content-generator-job` retries to get the
+  collections from collection cache."
+  {:default 20
+   :type Long})
+
+(defconfig generation-interval
+  "Number of seconds between `static-content-generator-job` runs."
+  {;:default (* 60 60 24)
+   :default (* 60)
+   :type Long})
+
+(defn create-cache
+  "This function creates the composite cache that is used for caching the
+  static content data. With the given composition we get the following
+  features:
+  * A cubby cache that holds the generated content (centralized storage in
+    an ElasticSearch backend);
+  * A single-threaded cache that circumvents potential race conditions
+    between HTTP requests for a report and Quartz cluster jobs that save
+    report data.
+
+  This function is intended to be called by a function that sets up caches for
+  an application (e.g., cmr.search.system/create-system)."
+  []
+  (stl-cache/create-single-thread-lookup-cache
+   (cubby-cache/create-cubby-cache)))
+
+(defn get-page-content
+  "Given a context and route (optionally provider id and tag), get the
+  output of the page template that corresponds to the route."
+  ([context route]
+   (case route
+     "/sitemap.xml" (pages/sitemap-master context)
+     "/site/sitemap.xml" (pages/sitemap-top-level context)
+     "/site/collections/directory/eosdis" (pages/eosdis-collections-directory context)))
+  ([context route provider-id tag]
+   (cond
+     (string/ends-with? route ".xml")
+     (pages/sitemap-provider-tag context provider-id tag)
+     :else (pages/provider-tag-directory context provider-id tag))))
+
+(defn cache-page
+  "Cache a page for a given route. If the content is supplied, simply cache
+  that; if only a route is supplied, get the content for that route and then
+  cache it. If a provider id and tag are supplied, get the appopriate content
+  given those args, and cache the result."
+  ([context route]
+   (cache-page context route (get-page-content context route)))
+  ([context route content]
+   (debug "Caching" route)
+   (cache/set-value (cache/context->cache context cache-key) route content))
+  ([context route provider-id tag]
+    (cache-page context
+                route
+                (get-page-content context route provider-id tag))))
+
+(defn retrieve-page
+  "Attempt to retrieve from the cache the page data for a given route. If no
+  value is found in the cache, genereate the content, cache it, and then return
+  the cached value."
+  [& args]
+    (let [context (first args)
+          route (second args)]
+      (debug "Retrieving" route)
+      (if-let [content (cache/get-value (cache/context->cache context cache-key)
```

Comment:
You'll want to use a lookup-fn which gets used when the particular page is not cached rather than using an if-else. The cache will automatically save the value returned by the lookup-fn in the cache.

```
  (get-value
    [cache key]
    [cache key lookup-fn]
    "Looks up the value of the cached item using the key. If there is a cache miss
    and a function is provided it will invoke the function given with no arguments, save the
    returned value in the cache, then return the value. If no function is provided it will return nil
    for cache misses.")
```

Reply:
Oh, nice -- thanks. Wasn't aware of that feature.

Done.